x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -145,9 +189,67 @@ export const performBulkActionRoute = (
               },
               body: responseBody,
             });
+          case BulkAction.edit:
+            processingResponse = await executeBulkAction(rules.data, async (rule) => {
+              throwHttpError(await mlAuthz.validateRuleType(rule.params.type));
```

Comment:
@banderror , looks like there is already implemented optimization for ML check
https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/machine_learning/authz.ts#L44
so, it should call ML services only once, on mlAuthz initialization and then return cached value when calling `mlAuthz.validateRuleType`. 

Are you still keen to optimize further? How do you see then this optimization?

From top of my head:

First approach
1. store result of ML rule validation for ML type of rule as a separate value
2. when processing action, throw error when type is ML

we still would need to check in every action whether rule is ML, we'll need another function for that to wrap it, if we don't want just put there if clause. Essentially will have N calls for that new function or if statement

Second approach
1. Loop over all found rules firstly, and throw error if it is ML and fails validation, gather caught errors
2. Proceed with bulk action in another loop

In this case we will add another loop through rules. Again will have additional N calls and a bit more complicated logic

What do you think? Anything we can improve here?

Reply:
Discussed over zoom.
We don't need to optimize the performance because this call is cached, but it would make sense to normalize the response in such a way so that duplicate errors are de-duplicated, something like that:

```ts
          const responseBody = {
            message: errorsCount === rulesCount ? 'Bulk edit failed' : 'Bulk edit partially failed',
            status_code: 500,
            attributes: {
              errors: errors.map(({ error, rule }) => ({
                message:
                  error.message.length > MAX_ERROR_MESSAGE_LENGTH
                    ? error.message.slice(0, MAX_ERROR_MESSAGE_LENGTH)
                    : error.message,
                status_code: error.statusCode,
                rules: [{ id, name }]
              })),
              rules: {
                total: rulesCount,
                failed: errorsCount,
                succeeded: rulesCount - errorsCount,
              },
            },
          };
```